### DISCUSSION TITLE
question

### DISCUSSION THREAD

**Discussion by Chang-Zhang** [ss3sim/ss3sim#350] [Q&A]
When running the example data and code, it showns: Error in `[.data.frame`(baseom_tv, , c("env-var", "use_dev", "Block")). Where should I change it?

code:
d <- system.file("extdata", package = "ss3sim")
case_folder <- file.path(d, "eg-cases")#index0-cod.txt, E0-cod.txt; F0-cod.txt; M0-cod.txt
om <- file.path(d, "models", "cod-om")
em <- file.path(d, "models", "cod-em")
run_ss3sim(iterations = 1, scenarios =
             c("D0-E0-F0-M0-cod",
               "D1-E0-F0-M0-cod",
               "D0-E1-F0-M0-cod",
               "D1-E1-F0-M0-cod"),
           case_files = list(F = "F", D = c("index", "lcomp", "agecomp"),
                             E = "E", M = "M"),
           case_folder = case_folder, om_dir = om,
           em_dir = em)

**Comment by kellijohnson-NOAA** [comment]
Thank you @Chang-Zhang for you question regarding running ss3sim. The use of case files is no longer recommended, and instead, we recommend using the simulation data frame. The [Introduction vignette](https://ss3sim.github.io/ss3sim/articles/introduction.html) has some information on [how to set up the simulation data frame](https://ss3sim.github.io/ss3sim/articles/introduction.html#scenarios). Can I ask where this example is in the package so I can update it with the newest best practices?
Also, I recommend updating to the newest version of the package, where if you run your code you should get
```
Error: The `scenarios` argument of `run_ss3sim()` was deprecated in ss3sim 1.1.4 and is now defunct.
```

**Reply by Chang-Zhang** [reply]
The example I used is from https://cran.r-project.org/web/packages/ss3sim/vignettes/introduction.html#example.
I still have some problems, 1. it shows could not find function "setup_scenarios_defaults", after library(ss3sim) and library(r4ss);
2. after creating new files of OM and EM, how to read this files?
3.I want to use different files between OM and EM(OM with time-varying, EM without time-varying), how to start simulations.

**Comment by kellijohnson-NOAA** [comment]
Ahh thank you for providing the link. The CRAN version of ss3sim is not up to date. You can install the latest GitHub version using the following code
```
library(devtools)
unload(package = "ss3sim")
unload(package = "r4ss")
remotes::install_github("r4ss/r4ss")
remotes::install_github("ss3sim/ss3sim")
library(ss3sim)
```

**Reply by Chang-Zhang** [reply]
Thank you for the answer. One more question, when I use the my own OM and EM files, "df<- setup_scenarios_defaults()" does not read the new file, the data frame of "df" is the same as in the example. I want to know it is a right way to read new files.
 
df<- setup_scenarios_defaults()
df
 cf.years.1      cf.fvals.1           si.years.2 si.sds_obs.2 si.seas.2 sl.Nsamp.1
1     26:100 rep(0.1052, 75) seq(62, 100, by = 2)          0.1         1         50
  sl.years.1 sl.Nsamp.2           sl.years.2 sl.cpar sa.Nsamp.1 sa.years.1 sa.Nsamp.2
1     26:100        100 seq(62, 100, by = 2)    NULL         50     26:100        100
            sa.years.2 sa.cpar
1 seq(62, 100, by = 2)    NULL

**Comment by kellijohnson-NOAA** [comment]
A column should be added to df for the OM called `om_dir` and another column for the EM dir called `em_dir`.

**Reply by Chang-Zhang** [reply]
d<-system.file("extdata",package="ss3sim")
om<-file.path(d,"models","om")
em<-file.path(d,"models","em")
iterations <- 1:5
scname <- run_ss3sim(iterations = iterations, simdf = df, om_dir =om,em_dir =em)
"Error: The `om_dir` argument of `run_ss3sim()` was deprecated in ss3sim 1.1.4 and is now defunct."

Is it a right way to add columns?

**Reply by k-doering-NOAA** [reply]
"om_dir" and "em_dir" should be columns within df (example: https://github.com/k-doering-NOAA/cal-sims/blob/ce4ada6df484e8b941d81227e7e4419f051569ad/code/piner_et_al_scenarios.R#L89-L90)

**Reply by Chang-Zhang** [reply]
There are two possible problems with this code:
line 47:[1] FALSE,Warning message:In r4ss::copy_SS_inputs(dir.old = em_init_path, dir.new = file.path(mod_files_path,  :at least 1 file failed to copy
line140Error in run_analysis(iter_vec = 1:10, simdf = df, results_wd = out) : could not find function "run_analysis".

**Reply by k-doering-NOAA** [reply]
This is the function run_analysis is referring to: https://github.com/k-doering-NOAA/cal-sims/blob/ce4ada6df484e8b941d81227e7e4419f051569ad/R/run_analysis.R

I believe the warning message can be ignored.

**Reply by Chang-Zhang** [reply]
it still cannot run.
Running scenarios and iterations sequentially.
Error in copy_ss3models(model_dir = om_dir, scenarios = sc, iterations = i,  : 
  argument "om_dir" is missing, with no default
In addition: Warning messages:
1: In if (.x[["label"]] == "user_recdevs") { :
  the condition has length > 1 and only the first element will be used
2: In if (.x[["label"]] == "user_recdevs") { :
  the condition has length > 1 and only the first element will be used
3: In if (.x[["label"]] == "user_recdevs") { :

**Reply by kellijohnson-NOAA** [reply]
I am sorry this has proved to be so difficult. I am happy to join a video conference to help work through these issues with you if you would like. You can send the invitation to kelli.johnson@noaa.gov if you wish.
